File: src/concurrency/issues/dword_atomicity.c
#include <pthread.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>

uint64_t global = 0;

static void check() {
    uint64_t doubleWord = global;
    if ((doubleWord >> 32) != (doubleWord & 0xffffffff)) {
        fprintf(stderr, "Concurrency issue detected\nRead: 0x%llx\n", doubleWord);
        exit(1);
    }
}

static void *runTest(void *voidPtr) {
    uint64_t doubleWord = *((uint64_t *) voidPtr);
    uint32_t i;
    for (i = 0; i < 10000000; i++) {
        global = doubleWord;
        check();
    }
    return NULL;
}

int main() {
    pthread_t threads[8];
    uint32_t i;
    uint64_t ids[8];
    for (i = 0; i < 8; i++) {
        ids[i] = ((uint64_t) i + 1) << 32 | i + 1;
        fprintf(stdout, "Created job with id: %llx\n", ids[i]);
        if (pthread_create(&threads[i], NULL, runTest, &ids[i])) {
            fprintf(stderr, "Error creating thread\n");
            return 1;
        }
    }

    for (i = 0; i < 8; i++) {
        if (pthread_join(threads[i], NULL)) {
            fprintf(stderr, "Failed to join on thread\n");
            return 1;
        }
    }

    printf("No problems detected\n");

    return 0;
}

File: src/concurrency/issues/DWordAtomicity.java
package concurrency.issues;

import static concurrency.util.RunnerUtil.create;
import static concurrency.util.RunnerUtil.join;
import static concurrency.util.RunnerUtil.start;

public class DWordAtomicity {
  private static final int ITERATIONS = 1000000;

  public static void main(String[] args) throws InterruptedException {
    State state = new State();

    Job[] jobs = new Job[8];
    for (int id = 0; id < jobs.length; id++) {
      jobs[id] = new Job(state, id + 1);
    }
    Thread[] threads = create(jobs);
    start(threads);
    join(threads);

    System.out.println("No problems detected");
  }

  private static class State {
    private long global;

    private static void validate(long theLong) {
      long high = theLong >>> 32;
      long low = 0xffffffffL & theLong;

      if (high != low) {
        System.err.println("Concurrency issue detected. Read: 0x" + Long.toHexString(theLong) + "L");
        System.exit(1);
      }
    }

    public void setGlobal(long id) {
      global = id;

      validate(global);
    }
  }

  private static class Job implements Runnable {
    private final State state;
    private final long id;

    private Job(State state, long number) {
      this.state = state;
      this.id = (number << 32) | number;

      System.out.println("Created job with id: 0x" + Long.toHexString(id) + "L");
    }

    public void run() {
      for (int i = 0; i < ITERATIONS; i++) {
        state.setGlobal(id);
      }
    }
  }
}
